src/commontransformations.jl

```diff
@@ -1,41 +1,283 @@
+"""
+    abstract AbstractAffineTransformation <: Transformation
+
+Provides an interface for implementing Affine transformations of Cartesian
+coordinates. To implement an AbstractAffineTransformation, you must define
+
+    transformation_matrix(trans)
+    translation_vector(trans)
+
+where the resulting transformation is (equivalent to)
+
+    trans(x) -> transformation_matrix(trans) * x + translation_vector(trans)
+
+Specific implementations may provide equivalent specializations of `call`, etc,
+for optimization purposes. The function `translation_vector_reverse()` is provided,
+such that
+
+    trans(x) -> transformation_matrix(trans) * (x + translation_vector_reverse(trans))
+
+(See also AffineTransformation, AbstractLinearTransformation, Translation)
+"""
+abstract AbstractAffineTransformation <: Transformation
+
+transformation_matrix(trans::AbstractAffineTransformation) = error("AbstractAffineTransformation $(typeof(trans)) must implement transformation_matrix()")
+translation_vector(trans::AbstractAffineTransformation) = error("AbstractAffineTransformation $(typeof(trans)) must implement translation_vector()")
+translation_vector_reverse(::AbstractAffineTransformation) = transformation_matrix(trans) \ translation_vector(trans)
+
+# Default implementations
+@compat function (trans::AbstractAffineTransformation)(x)
+    transformation_matrix(trans) * x + translation_vector(trans)
+end
+
+# Could try do similar for transform_deriv_params()?
+
+transform_deriv(trans::AbstractAffineTransformation, x) = transformation_matrix(trans)
+
+function Base.inv(trans::AbstractAffineTransformation)
+    Minv = inv(transformation_matrix(trans))
+    AffineTransformation(Minv, -Minv * translation_vector(trans))
+end
+
+function Base.isapprox(t1::AbstractAffineTransformation, t2::AbstractAffineTransformation, kwargs...)
+    isapprox(transformation_matrix(t1), transformation_matrix(t2); kwargs...) &&
+        isapprox(translation_vector(t1), translation_vector(t2); kwargs...)
+end
+
+"""
+    abstract AbstractLinearTransformation <: AbstractAffineTransformation
+
+Provides an interface for implementing linear transformations of Cartesian
+coordinates. To implement an AbstractLinearTransformation, you must define
+
+    transformation_matrix(trans)
+
+where the resulting transformation is (equivalent to)
+
+    trans(x) -> transformation_matrix(trans) * x
+
+Specific implementations may provide equivalent specializations of `call`, etc,
+for optimization purposes.
+
+(See also LinearTransformation, AbstractAffineTransformation)
+"""
+abstract AbstractLinearTransformation <: AbstractAffineTransformation
+
+transformation_matrix(trans::AbstractLinearTransformation) = error("AbstractLinearTransformation $(typeof(trans)) must implement transformation_matrix()")
+function translation_vector(trans::AbstractLinearTransformation)
+    m = transformation_matrix(trans)
+    s = size(m, 1)
+    T = eltype(m)
+    return zeros(T, s)
```

Comment:
This isn't ideal for use with FixedSizeArrays.  Maybe it doesn't matter because calling `translation_vector` on a LinearTransformation isn't really something that makes a whole lot of sense?